### PR TITLE
Use a chain pattern for data providers

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #506
| License       | MIT
| Doc PR        | todo

Use a chain instead of a decorator for data providers. It allows to register custom data providers in a more intuitive fashion: just add a tag to your service.

ping @soyuka

### CODE REVIEW

**Comment by soyuka** [comment]
Nice! Thx ❤️ .

File: src/DataProvider/ChainCollectionDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class ChainCollectionDataProvider implements CollectionDataProviderInterface
+{
+    private $collectionDataProviders;
+
+    /**
+     * @param CollectionDataProviderInterface[] $collectionDataProviders
+     */
+    public function __construct(array $collectionDataProviders)
```
Comment (theofidry) [inline src/DataProvider/ChainCollectionDataProvider.php]: `$dataProviders`?

File: src/DataProvider/ChainCollectionDataProvider.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+namespace ApiPlatform\Core\DataProvider;
+
+use ApiPlatform\Core\Exception\ResourceClassNotSupportedException;
+
+/**
+ * Tries each configured data provider and returns the result of the first able to handle the resource class.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class ChainCollectionDataProvider implements CollectionDataProviderInterface
```
Comment (theofidry) [inline src/DataProvider/ChainCollectionDataProvider.php]: as it implements an interface, what about making it `final`?

File: src/Bridge/Doctrine/Orm/CollectionDataProvider.php [hunk 3]
```diff
@@ -27,33 +27,22 @@ class CollectionDataProvider implements CollectionDataProviderInterface
...
     private $collectionExtensions;
-    private $decorated;
 
     /**
      * @param ManagerRegistry                      $managerRegistry
      * @param QueryCollectionExtensionInterface[]  $collectionExtensions
-     * @param CollectionDataProviderInterface|null $decorated
      */
-    public function __construct(ManagerRegistry $managerRegistry, array $collectionExtensions = [], CollectionDataProviderInterface $decorated = null)
+    public function __construct(ManagerRegistry $managerRegistry, array $collectionExtensions = [])
```
Comment (theofidry) [inline src/Bridge/Doctrine/Orm/CollectionDataProvider.php]: @dunglas this one could be final as well

**Comment by theofidry** [comment]
👍

**Pull request merged**